tests/test_options.py

```diff
@@ -1455,31 +1688,252 @@ def cli(one, two):
         runner.invoke(cli, [])
 
 
+def test_custom_type_click_class_flag_value(runner):
+    """A reproduction of
+    https://github.com/pallets/click/issues/3024#issuecomment-3146511356
+    """
+
+    NO_CONFIG = object()
+
+    class ConfigParamType(click.ParamType):
+        name = "config"
+
+        def convert(self, value, param, ctx):
+            if value is NO_CONFIG:
+                return value
+            else:
+                return click.Path(exists=True, dir_okay=False).convert(
+                    value, param, ctx
+                )
+
+    @click.command()
+    @click.option("-c", "--config", type=ConfigParamType())
+    @click.option("--no-config", "config", flag_value=NO_CONFIG, type=ConfigParamType())
+    def main(config):
+        click.echo(repr(config), nl=False)
+
+    result = runner.invoke(main)
+    assert result.output == repr(None)
+
+
+class EngineType(enum.Enum):
+    OSS = enum.auto()
+    PRO = enum.auto()
+    MAX = enum.auto()
+
+
+class Class1:
+    pass
+
+
+class Class2:
+    pass
+
+
 @pytest.mark.parametrize(
-    ("multiple", "nargs", "default", "expected_message"),
+    ("opt_params", "args", "expected"),
     [
+        # Reproduction of cases from
+        # https://github.com/pallets/click/issues/3024#issuecomment-3146480714
```

Comment:
Describe what's being tested, not only a link.

Reply:
Fixed in https://github.com/pallets/click/pull/3030/commits/354788eaba85d2ef405b300186a3752c2bc4dbaa